python-checks/src/main/java/org/sonar/python/checks/ProvideCorrectParameterDateTimeConstructorsCheck.java

```diff
@@ -0,0 +1,117 @@
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.NumericLiteral;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.plugins.python.api.tree.UnaryExpression;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6882")
+public class ProvideCorrectParameterDateTimeConstructorsCheck extends PythonSubscriptionCheck {
+  private static final int MIN_YEAR = 1;
+  private static final int MAX_YEAR = 9999;
+  private static final String MESSAGE = "The %s parameter must be valid.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, ProvideCorrectParameterDateTimeConstructorsCheck::checkCallExpr);
+  }
+
+  private static void checkCallExpr(SubscriptionContext context) {
+    CallExpression callExpression = (CallExpression) context.syntaxNode();
+    Symbol calleeSymbol = callExpression.calleeSymbol();
+    if (calleeSymbol == null) {
+      return;
+    }
+    if ("datetime.date".equals(calleeSymbol.fullyQualifiedName())) {
+      checkDate(context, callExpression);
+    } else if ("datetime.time".equals(calleeSymbol.fullyQualifiedName())) {
+      checkTime(context, callExpression);
+    } else if ("datetime.datetime".equals(calleeSymbol.fullyQualifiedName())) {
+      checkDate(context, callExpression);
+      checkTime(context, callExpression, 3);
+    }
+  }
+
+  private static void checkTime(SubscriptionContext context, CallExpression callExpression) {
+    checkTime(context, callExpression, 0);
+  }
+
+  private static void checkTime(SubscriptionContext context, CallExpression callExpression, int parameterOffset) {
+    RegularArgument hourArgument = TreeUtils.nthArgumentOrKeyword(parameterOffset, "hour", callExpression.arguments());
+    RegularArgument minuteArgument = TreeUtils.nthArgumentOrKeyword(parameterOffset + 1, "minute", callExpression.arguments());
+    RegularArgument secondArgument = TreeUtils.nthArgumentOrKeyword(parameterOffset + 2, "second", callExpression.arguments());
+    RegularArgument microsecondArgument = TreeUtils.nthArgumentOrKeyword(parameterOffset + 3, "microsecond", callExpression.arguments());
+
+    if (hourArgument != null) {
+      checkArgument(hourArgument, 0, 23, context, "hour");
+    }
+    if (minuteArgument != null) {
+      checkArgument(minuteArgument, 0, 59, context, "minute");
+    }
+    if (secondArgument != null) {
+      checkArgument(secondArgument, 0, 59, context, "second");
+    }
+    if (microsecondArgument != null) {
+      checkArgument(microsecondArgument, 0, 999_999, context, "microsecond");
+    }
+  }
+
+  private static void checkArgument(RegularArgument argument, long min, long max, SubscriptionContext context, String name) {
+    if (!argument.expression().is(Tree.Kind.NUMERIC_LITERAL) && !argument.expression().is(Tree.Kind.UNARY_MINUS)) {
```

Comment:
You can probably cover some more cases by calling `Expressions#singleAssignedNonNameValue` or `Expressions#singleAssignedValue` without too much overhead when the argument expression is a `NAME`.
If you decide to cover this case, it may make sense to have a primary location on the argument itself and a secondary on the value with message such as e.g `"An invalid value is assigned here."`